framework/src/utils/MooseCoordTransform.C

```diff
@@ -0,0 +1,338 @@
+      "angle ZXZ convention from https://en.wikipedia.org/wiki/Euler_angles#Rotation_matrix in "
+      "order to align with a canonical physical space of your choosing.");
+  MooseEnum up_direction("X=0 Y=1 Z=2");
+  params.addParam<MooseEnum>(
+      "up_direction",
+      up_direction,
+      "Specify what axis corresponds to the up direction in physical space (the opposite of the "
+      "gravity vector if you will). If this parameter is provided, we will perform a single 90 "
+      "degree rotation of the domain--if the provided axis is 'x' or 'z', we will not rotate if "
+      "the axis is 'y'--such that a point which was on the provided axis will now lie on the "
+      "y-axis, e.g. the y-axis is our canonical up direction. If you want finer grained control "
+      "than this, please use the 'alpha_rotation', 'beta_rotation', and 'gamma_rotation' "
+      "parameters.");
+  return params;
+}
+
+MooseCoordTransform::MooseCoordTransform(const InputParameters & params)
+{
+  //
+  // coordinate system transformation. If we have multiple different coordinate system types in our
+  // problem, we take note of it because that can cause issues if there is a non-Cartesian
+  // destination coordinate system
+  //
+  const auto & mesh = *params.getCheckedPointerParam<MooseMesh *>("mesh");
+  const auto & coord_sys = mesh.getCoordSystem();
+  std::unordered_set<Moose::CoordinateSystemType> coord_types;
+  auto map_it = coord_sys.begin();
+  // It's possible that the mesh is not in a complete state
+  if (map_it == coord_sys.end())
+    setCoordinateSystem(Moose::COORD_XYZ);
+  else
+    setCoordinateSystem(
+        map_it->second,
+        Direction(static_cast<unsigned int>(int(params.get<MooseEnum>("rz_coord_axis")))));
+  for (; map_it != coord_sys.end(); ++map_it)
+    coord_types.insert(map_it->second);
+
+  _has_different_coord_sys = coord_types.size() > 1;
+
+  //
+  // rotation
+  //
+  const bool has_alpha = params.isParamValid("alpha_rotation");
+  const bool has_beta = params.isParamValid("beta_rotation");
+  const bool has_gamma = params.isParamValid("gamma_rotation");
+  const auto & up_direction = params.get<MooseEnum>("up_direction");
+
+  if (has_alpha || has_beta || has_gamma)
+  {
+    if (up_direction.isValid())
+      mooseError("Cannot simultaneously set rotation angles as well as an up direction");
+
+    const auto alpha = (has_alpha ? params.get<Real>("alpha_rotation") : Real(0));
+    const auto beta = (has_beta ? params.get<Real>("beta_rotation") : Real(0));
+    const auto gamma = (has_gamma ? params.get<Real>("gamma_rotation") : Real(0));
+
+    setRotation(alpha, beta, gamma);
+  }
+  else if (up_direction.isValid())
+    setUpDirection(Direction(static_cast<unsigned int>(int(up_direction))));
+
+  //
+  // Scale
+  //
+  if (params.isParamValid("length_units_per_meter"))
+    setLengthUnitsPerMeter(params.get<Real>("length_units_per_meter"));
```

Comment:
```suggestion
    setLengthUnitsPerMeter(Real(MooseUnits(params.get<Real>("length_units_per_meter"))/MooseUnits("m")));
```

Reply:
I would suggest setting this to be ratio between _our_ length unit and the _destination_ length unit.

Reply:
Also, storing a publicly accessible member `MooseUnits _length_unit` should satisfy @YaqiWang 's needs.

Reply:
The downside is that users may no longer specify arbitrary scaling factors. Perhaps we should take either a `std::string` for MooseUnits or a `Real`.

Reply:
You can always specify arbitrary scaling factors. `3.89in` is a valid unit...

Reply:
Fantastic!

Reply:
> I would suggest setting this to be ratio between our length unit and the destination length unit

wherever possible I would like each application's "transformation" information to be as encapsulated as possible. E.g. I would like it's data (as much as possible) to only reflect itself. Because you may have different ratios for example when transferring to/from/between different sub-application. For the "coordinate collapsing" case, that necessitates holding some destination information, which we've already seen can cause confusion ... so if I can limit the potential sources of confusion, I would like to

Reply:
I agree with you. I didn't consider the case of multiple coord transformations across different apps.

Reply:
@dschwen I'm seeing '20cm' get parsed into
```
(gdb) p _length_unit
$2 = {_factor = 0.01, _shift = 0, _base = std::vector of length 1, capacity 1 = {{
      first = MooseUnits::BaseUnit::METER, second = 1}}}
```
It seems like the '20' prefactor is getting ignored

Reply:
'20*cm' gets parsed correctly